### PR TITLE
Introduce 4 static memory allocation policies

### PR DESCRIPTION
**Pull request by IoanAnev**
This pull request introduces 4 static memory allocation policies, namely _cyclic_, _skew-mapp_, _prime-mapp_, and _first-touch_. The implementations can address different granularities rather than only handling data at the default page granularity level (4KB).

The memory allocation policy is selected through the environment variable `ARGO_ALLOCATION_POLICY`, which should be given a number from **0** to **4**, starting from the default naive distribution and continuing with the rest of the policies in the order presented above. The page block size can be adjusted by setting `ARGO_ALLOCATION_BLOCK_SIZE`. In case the former environment variable is not specified, the _naive_ distribution will be used and, in the case of the latter, a size of **16** will be selected, resulting in a granularity of 16*4KB=64KB.

### CODE REVIEW

**Comment by lundgren87** [comment]
### Issue
As I informed @IoanAnev of last week, there is a potential deadlock situation when performing writeback (release or acquire with dirty pages) on pages that are not strictly ordered in respect of homenode (`0, 1, ..., n` OR `n, n-1, ..., 0`). If not ordered, the following situation may appear:
deadlock

With cyclic or first touch allocation policies, this ordering should be fairly common and I believe that this has not been immediately caught by testing and benchmarking as it only appears to occur frequently with four nodes (can not occur with two, possible not with three) and on larger memory sizes. It is still possible for any workload to pass by pure coincidence.

### Solution
In native ArgoDSM this is avoided by iterating over the cache during acquire and release, guaranteeing due to the native allocation policy that remote nodes will be locked in order as they are already ordered in the cache. In #45, this is instead achieved by sorting the write buffer in order of cache index before performing a writeback.

The solution in #45 could easily be extended to sorting the write buffer in order of homenode before performing the writeback. The reasoning for sorting only before writeback is to keep the FIFO ordering of the buffer until a synchronization point, so that the least recently pages are written back if the write buffer becomes full between synchronization points.
One would have to consider the performance implications of _first touch_ when querying homenode for many pages in succession, but these operations would be performed sooner or later and it may not make any difference to do these operations all at once. In fact, there is possibly potential for optimizing the homenode queries by finding the homenode of multiple pages in one remote call.

File: src/data_distribution/first_touch_distribution.hpp
Comment (lundgren87) [inline src/data_distribution/first_touch_distribution.hpp:98]: Unclear what this index is, could use a more descriptive name. Suggesting `owner_window_index` (if `ownerWindow` is renamed to `owner_window`).

File: src/data_distribution/first_touch_distribution.hpp
Comment (lundgren87) [inline src/data_distribution/first_touch_distribution.hpp:105]: It may be worth being careful here (in general for all distributions) to avoid confusion, `homenode` in most places of the backend refers to the ID returned by `argo::node_id()` (and equivalently `getHomenode(...)`, `getID()` and `workrank` in the MPI backend) which is in integer format (0 to n-1 for n nodes), while `id` here is the "bit representation". I suggest using a specific name for the bit representation such as `homenode_id` or maybe even `homenode_bit_id` when referring to the bit representation, and `homenode` when referring to the plain integer representation.

File: src/data_distribution/naive_distribution.hpp
Comment (lundgren87) [inline src/data_distribution/naive_distribution.hpp:15]: As mentioned before, there should be a better place for these rather than in a specific distribution implementation.

File: src/data_distribution/prime_mapp_distribution.hpp
Comment (lundgren87) [inline src/data_distribution/prime_mapp_distribution.hpp:21]: This is confusing as it does not return a prime number for most number of nodes. Need to figure out what the limitations are as per our discussions and either document this and prevent execution with the wrong number of nodes, or delay the inclusion of the prime_mapp distributions in this PR.

**Review by lundgren87 — CHANGES_REQUESTED** [review]
Overall looks okay, still pending thorough testing on Jenkins after merging #45 and rebasing on top of it in order to fix the deadlocking bug mentioned in the comments.

**Big things to fix:**
- Definition of the base distribution class in its own file and adjustments accordingly (suggestion mentioned in specific comment).
- Determine the limitations of the prime mapp policy and document plus limit/disable execution for unsupported node numbers, or alternatively leave this distribution out if there is no clear answer.

**Minor things to fix:**
- snake_case for all variables, functions and classes.
- datatypes and variable names according to comments
- a few indentations slipped through

File: src/backend/mpi/mpi.cpp
Comment (IoanAnev) [reply src/backend/mpi/mpi.cpp]: Changed as requested.

File: src/backend/mpi/swdsm.cpp
Comment (IoanAnev) [reply src/backend/mpi/swdsm.cpp]: Changed as requested.

File: src/backend/mpi/swdsm.cpp
Comment (IoanAnev) [reply src/backend/mpi/swdsm.cpp]: Changed as requested.
Introduced a `offset` variable of type `std::size_t` in that case.

File: src/backend/mpi/swdsm.cpp
Comment (IoanAnev) [reply src/backend/mpi/swdsm.cpp]: Changed as requested.
Introduced a `homenode` variable of type `std::size_t` in that case.

File: src/backend/mpi/swdsm.cpp
Comment (IoanAnev) [reply src/backend/mpi/swdsm.cpp]: Changed as requested.

File: src/backend/mpi/swdsm.cpp
Comment (IoanAnev) [reply src/backend/mpi/swdsm.cpp]: Changed as requested.

File: src/backend/mpi/swdsm.cpp
Comment (IoanAnev) [reply src/backend/mpi/swdsm.cpp]: Changed as requested.

File: src/backend/mpi/swdsm.cpp
Comment (IoanAnev) [reply src/backend/mpi/swdsm.cpp]: Changed as requested.

File: src/backend/mpi/swdsm.h
Comment (IoanAnev) [reply src/backend/mpi/swdsm.h]: Changed as requested.

File: src/backend/singlenode/singlenode.cpp
Comment (IoanAnev) [reply src/backend/singlenode/singlenode.cpp:98]: No change was made.

File: src/data_distribution/cyclic_distribution.hpp
Comment (IoanAnev) [reply src/data_distribution/cyclic_distribution.hpp]: Changed as requested.

File: src/data_distribution/cyclic_distribution.hpp
Comment (IoanAnev) [reply src/data_distribution/cyclic_distribution.hpp]: Changed as requested.

File: src/data_distribution/cyclic_distribution.hpp
Comment (IoanAnev) [reply src/data_distribution/cyclic_distribution.hpp:36]: No changes were made.
The _naive-distribution_ handles unaligned addresses by default.

File: src/data_distribution/data_distribution.hpp
Comment (IoanAnev) [reply src/data_distribution/data_distribution.hpp]: Changed as requested.

File: src/data_distribution/first_touch_distribution.hpp
Comment (IoanAnev) [reply src/data_distribution/first_touch_distribution.hpp:34]: Changed as requested with a few modifications.
As being documented above the forward declarations in the relevant file, we cannot avoid the forward declarations here as backend.hpp includes global_ptr.hpp, which means that the distribution definitions precede the backend definitions. For that reason, the inclusion of `backend.hpp` under `first_touch_distribution.hpp` deemed to be reduntant and was removed.

File: src/data_distribution/first_touch_distribution.hpp
Comment (IoanAnev) [reply src/data_distribution/first_touch_distribution.hpp]: Changed as requested.

File: src/data_distribution/first_touch_distribution.hpp
Comment (IoanAnev) [reply src/data_distribution/first_touch_distribution.hpp]: Changed as requested.
We distinguish `homenode` from `homenode_bit_id`.

File: src/data_distribution/naive_distribution.hpp
Comment (IoanAnev) [reply src/data_distribution/naive_distribution.hpp]: Changes were made.
Both of the forward declarations are removed:
- The new file hierarchy removes the `base_distribution` forward declaration.
- The `global_ptr` forward declaration is removed by passing `base_distribution<0>` directly to the `global_ptr` class definition, as per the original format.

File: src/data_distribution/prime_mapp_distribution.hpp
Comment (IoanAnev) [reply src/data_distribution/prime_mapp_distribution.hpp:23]: The comment on the PR that the _prime-mapp_ policies work only for a specific number of nodes was misleading.
The bug was fixed some time ago, and the _prime-mapp_ policies work for any number of nodes.
Of course, Jenkins tests will justify this claim.

File: src/backend/mpi/swdsm.cpp
Comment (lundgren87) [reply src/backend/mpi/swdsm.cpp]: Re-visit this part as per our slack discussion to make sure it complies with master.

File: src/backend/singlenode/singlenode.cpp
Comment (lundgren87) [reply src/backend/singlenode/singlenode.cpp:98]: Ok, not not crucial.

File: src/data_distribution/prime_mapp_distribution.hpp
Comment (lundgren87) [reply src/data_distribution/prime_mapp_distribution.hpp:23]: Ok, this is fine as long as all tests and benchmarks pass.

File: src/data_distribution/first_touch_distribution.hpp
Comment (lundgren87) [reply src/data_distribution/first_touch_distribution.hpp:34]: It looks better now in my opinion. I will ask @davidklaftenegger for a second opinion if and when he has time, but this is a vast improvement on the first iteration.

File: src/data_distribution/naive_distribution.hpp
Comment (lundgren87) [reply src/data_distribution/naive_distribution.hpp]: Same as above here, unless @davidklaftenegger has any comments I will say that this is okay for now.

**Comment by lundgren87** [comment]
Looks better now, see the two comments I made in case David has any input. Anything that is not related to functionality can possibly be left as future work/issues.

Need to wait for #45 to be merged in order to fix the bug and then proceed to test on Jenkins ofcourse.

File: src/data_distribution/first_touch_distribution.hpp
Comment (lundgren87) [inline src/data_distribution/first_touch_distribution.hpp:54]: I did not spot this in my initial review, but we should probably attempt to use `std::mutex` (combined with `std::lock_guard` or `std::unique_lock`) in new code instead of the pthread equivalent for portability and future-proofing (and RAII even if that may not be applicable here).

**Review by lundgren87 — CHANGES_REQUESTED** [review]
Two minor things I suggest fixing before we run on Jenkins noted in the comments below, otherwise it now looks okay.

**Review by lundgren87 — APPROVED** [review]
This patch now looks fine, Jenkins tests pass and benchmarks show great improvements overall for all new policies.

**Pull request merged**
